File: Python3/637. Average of Levels in Binary Tree.py
# 60ms, 65.82%; 16.2MB, 7.14%
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def averageOfLevels(self, root: TreeNode) -> list:
        res, st = [], [root]
        while st:
            res.append(0)
            tmp = []
            for node in st:
                res[-1] += node.val
                if node.left: tmp.append(node.left)
                if node.right: tmp.append(node.right)
            res[-1] /= len(st)
            st = tmp
        return res

File: 601 - 800/637. Average of Levels in Binary Tree.cc
// 16ms, 91.39%; 21.9MB, 44.44%
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<double> averageOfLevels(TreeNode *root) {
        vector<double> res;
        vector<TreeNode *> st;
        st.push_back(root);
        while (!st.empty()) {
            res.push_back(0);
            vector<TreeNode *> tmp;
            for (auto *node : st) {
                res.back() += node->val;
                if (node->left) tmp.push_back(node->left);
                if (node->right) tmp.push_back(node->right);
            }
            res.back() /= st.size();
            st.swap(tmp);
        }
        return res;
    }
};
